rfcs/45-stageless.md

```diff
@@ -0,0 +1,727 @@
+Typically, these are used for relatively complex, far-reaching facets of the game like pausing, loading assets or entering into menus.
+
+The current value of each state is stored in a resource, which must derive the `State` trait.
+These are typically (but not necessarily) enums, where each distinct state is represented as an enum variant.
+
+Each state is associated with three groups of systems:
+
+1. **While-active:** these systems run each schedule iteration if and only if the value of the state resource matches the provided value.
+   1. `app.add_system(apply_damage.run_if(state_equals(GameState::Playing))`
+2. **On-enter systems:** these systems run once when the specified state is entered.
+   1. `app.add_system(generate_map.in_set(OnEnter(GameState::Playing))`
+3. **On-exit systems:** these systems run once when the specified state is exited.
+   1. `app.add_system(autosave.in_set(OnExit(GameState::Playing)))`
+
+While-active systems are by far the simplest: they're simply powered by run criteria.
+
+On-enter and on-exit systems are stored in dedicated schedules, two per state, within the `App's` `Schedules`.
+These schedules can be configured in all of the ordinary ways, but, as they live in different schedules, ordering cannot be defined relative to systems in the main schedule.
+
+You can change which variant of a state you are in by requesting the state as a mutable resource, then using the `State::transition_to` method within your system.
+Due to their disruptive and far-reaching effects, state transitions do not occur immediately.
+Instead, they are deferred (like commands), until the next `flush_state<S: State>` exclusive system runs.
+This system first runs the `on_exit` schedule of the previous state on the world, then runs the `on_enter` schedule of the new state on the world.
+Once that is complete, the exclusive system ends and control flow resumes as normal.
+Note that commands are not automatically flushed between state transitions. If this is required, add the `flush_commands` system to your schedule.
+
+When states are added using `App::add_state::<S: State>(initial_state)`, one `flush_state<S>` system is added to the app, as part of the `GeneratedSet::StateTransition<S>` set.
+You can configure when and if this system is scheduled by configuring this set, and you can add additional copies of this system to your schedule where you see fit.
+
+Apps can have multiple orthogonal states representing independent facets of your game: these operate fully independently.
+States can also be defined as a nested enum: these work as you may expect, with each leaf node representing a distinct group of systems.
+
+### Flushing `Commands`
+
+Commands (commonly used to spawn and despawn entities or add and remove components) do not take effect immediately.
+Instead, they are applied whenever the `flush_commands` system runs.

```

Comment:
I think one single ungranular command flush deserves some scrutiny.

It means that we might apply commands "earlier than we need to". Given that the sum of all critical sections in theory will be the same no matter how you divide them up, this seems ok from a perf perspective (although in practice there might be real perf tradeoffs here).

But more importantly the non-granularity will also force some effects to be observed "earlier" than they need to be. Effects that don't have the need to be observed _at all_ this frame might now be observed before the majority of other systems this frame, just by happenstance according to how the scheduling plays out this frame. This could create situations where sometimes a system iterates a newly spawned entity this frame (as if it had a hard dependency on that effect) and other times it must wait until the next frame. 

Compare that to a scheduler that has a granular way to say "i want to observe the effects of this specific system". This would:
1. More accurately encode the intent of the system author / the requirements of the system (which imo is very important in this case).
2. Leave the door open to future optimizations (ex: I know we all would like to find a way to apply commands more granularly without a full lock).
3. Allow us to defer command application for as long as possible, ensuring more predictable behaviors. A system with no hard ordering constraints would have its commands applied as the last step in the schedule, meaning the effects would only be observed on the next update.

Reply:
I think we can afford to move incrementally here. Since we carve out command flushes into a separate primitive in this, imo this redesign already puts us in a better position than 0.7 for what you describe.

The points you make are very inline with the first set of follow-up RFCs highlighted in the last section (and the "side effect" stuff @hymm had been working on). We also wanted to have more nuanced relationships between systems than just `.before` and `.after`, but we removed those to reduce the scope of this RFC. Same with figuring out the optimal spots for flushes as a future optimization (that was "automatic sync point insertion"). 

> I know we all would like to find a way to apply commands more granularity than `&mut World`.

If an example of "more granular flushes" is "okay, just flush these 3 systems", that's definitely possible although I do expect performance tradeoffs since you'd probably end up with more flushes.

But I strongly doubt commands can avoid `&mut World` (at least not without creating a UI nightmare). Systems can mutate arbitrary archetypes, but more importantly they can spawn brand new ones. New archetypes would probably be matched by other systems, but they can't be known in advance, so you run into the same problem of "changes sometimes will and sometimes won't be detected on the same frame".